Mathlib/Topology/Connected/RealIntervals.lean

```diff
@@ -0,0 +1,560 @@
+/-
+Copyright (c) 2024 Dennis Sweeney. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dennis Sweeney
+-/
+import Mathlib.Data.Set.Card
+import Mathlib.Topology.MetricSpace.PseudoMetric
+
+/-!
+# Connected Subsets of ℝ
+
+In this file, we classify the connected subsets of ℝ.
+
+## Main Results
+
+  - `classify_connected_bounded_reals` : A bounded connected set of reals is one of:
+    - `Ioo a b` for some `a < b`,
+    - `Ioc a b` for some `a < b`,
+    - `Ico a b` for some `a < b`,
+    - `Icc a b` for some `a < b`,
+    - A singleton `{a}`.
+
+  - `classify_connected_bounded_reals_nonempty_interior` :
+    The same as above, but omitting the singleton case.
+
+  - `classify_connected_reals` : A connected set of reals is one of:
+    - The bounded connected sets of reals above,
+    - `Iio a` for some `a : ℝ`,
+    - `Iic a` for some `a : ℝ`,
+    - `Ioi a` for some `a : ℝ`,
+    - `Ici a` for some `a : ℝ`,
+    - `Set.univ : Set ℝ` (all of ℝ).
+
+  - `encard_frontier_BoundedInterval` : Bounded intervals have exactly two endpoints.
+
+## New Types
+
+  - `struct BoundedInterval` : Unifies nontrivial intervals `Ioo`, `Ioc`, `Ico`, and `Icc`.
+  - `inductive ConnectedRealClassification` : Unifies the various kinds of connected sets of reals.
+-/
+
+/-! ### Classifying Connected Bounded sets of Reals -/
+
+open Set
+
+
+lemma Ico_Ioo_with_endpoint {a b : ℝ} (hab : a < b) :
+    Ico a b = (Ioo a b) ∪ {a} := by
+  apply Subset.antisymm
+  · intro x ⟨x_le_a, x_lt_b⟩
+    by_cases xa : x = a
+    · right; exact xa
+    · left; exact ⟨Ne.lt_of_le' xa x_le_a, x_lt_b⟩
+  · rintro x (x_Ioo | x_a)
+    · exact ⟨LT.lt.le x_Ioo.1, x_Ioo.2⟩
+    · exact ⟨Eq.ge x_a, Eq.trans_lt x_a hab⟩
+
+lemma Ioc_Ioo_with_endpoint {a b : ℝ} (hab : a < b) :
+    Ioc a b = (Ioo a b) ∪ {b} := by
+  apply Subset.antisymm
+  · intro x ⟨a_lt_x, x_le_b⟩
+    by_cases xb : x = b
+    · right; exact xb
+    have : x < b := Ne.lt_of_le xb x_le_b
+    left; exact ⟨a_lt_x, this⟩
+  · rintro x (x_Ioo | x_b)
+    · exact ⟨x_Ioo.1, LT.lt.le x_Ioo.2⟩
+    · exact ⟨(by rw [x_b]; exact hab), Eq.le x_b⟩
+
+section X_section
+variable
+  {X : Set ℝ}
+
+lemma bdd_subset_Icc {infX supX : ℝ}
+    (h_infX : IsGLB X infX) (h_supX : IsLUB X supX) :
+    X ⊆ Icc infX supX := by
+  intro x xX
```

Comment:
This can be golfed on one line. Can you give `xX` a more descriptive name? (I have no idea what it is about. A hypothesis `x ∈ X` is often called `hx`.)